src/PhpSpec/Matcher/ArrayKeyValueMatcher.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+namespace PhpSpec\Matcher;
+
+use ArrayAccess;
+use PhpSpec\Exception\Example\FailureException;
+use PhpSpec\Formatter\Presenter\PresenterInterface;
+
+class ArrayKeyValueMatcher extends BasicMatcher
+{
+    /**
+     * @var \PhpSpec\Formatter\Presenter\PresenterInterface
+     */
+    private $presenter;
+
+    /**
+     * @param PresenterInterface $presenter
+     */
+    public function __construct(PresenterInterface $presenter)
+    {
+        $this->presenter = $presenter;
+    }
+
+    /**
+     * @param string $name
+     * @param mixed  $subject
+     * @param array  $arguments
+     *
+     * @return bool
+     */
+    public function supports($name, $subject, array $arguments)
+    {
+        return 'haveKeyWithValue' === $name
+        && 2 == count($arguments)
+        && (is_array($subject) || $subject instanceof ArrayAccess)
+            ;
+    }
+
+    /**
+     * @param ArrayAccess|array $subject
+     * @param array $arguments
+     *
+     * @return bool
+     */
+    protected function matches($subject, array $arguments)
+    {
+        $key = $arguments[0];
+        $value  = $arguments[1];
+
+        if ($subject instanceof ArrayAccess) {
+            return $subject->offsetExists($key) && $subject->offsetGet($key) === $value;
+        }
+
+        return (isset($subject[$key]) || array_key_exists($arguments[0], $subject) && $subject[$key] === $value);
+    }
+
+    /**
+     * @param string $name
+     * @param mixed  $subject
+     * @param array  $arguments
+     *
+     * @return FailureException
+     */
+    protected function getFailureException($name, $subject, array $arguments)
+    {
+        $key = $arguments[0];
+        $expectedValue = $arguments[1];
+        $actualValue = $subject[$key];
+
+        // Match failed because the key wasn't set
```

Comment:
please remove the comment

It's better to pull bits of this out into privates so that the overall shape is clear